### PR TITLE
Fix the incorrect permutation of gguf

### LINKED ISSUE #31766: The conversion of the llama3 model back from gguf seems weird.
### System Info

```
torch==2.3.1
accelerate==0.31.0
gguf==0.6.0
numpy==1.26.4
transformers==4.42.3
```

### Who can help?

@SunMarc

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

Prepare environment:
```sh
conda create -yn HF-GGUF python=3.11
conda activate HF-GGUF
conda install -y nvidia/label/cuda-12.1.1::cuda
conda install -y pytorch torchvision torchaudio pytorch-cuda=12.1 -c pytorch -c nvidia
pip install accelerate cmake datasets fire gguf hf-transfer "numpy<2.0" sentencepiece transformers
conda activate HF-GGUF
```

The two versions of the llama.cpp repository:
```sh
# tag: b1742, sha: edd1ab7, date: january
git clone --branch b1742 --single-branch https://github.com/ggerganov/llama.cpp llama-cpp-b1742
cmake -B build . -DLLAMA_CUBLAS=ON --fresh
cmake --build build --config Release -j

# tag: b3286, sha: fadde67, date: july
git clone --branch b3286 --single-branch https://github.com/ggerganov/llama.cpp llama-cpp-b3286
cmake -B build . -DGGML_CUDA=ON --fresh
cmake --build build --config Release -j
```

Prepare models:
```sh
# prepare hf models
HF_HUB_ENABLE_HF_TRANSFER=True huggingface-cli download meta-llama/Meta-Llama-3-8B --local-dir llama3-8b

# convert to gguf
python llama-cpp-b1742/convert.py --outtype f16 --outfile llama3-8b/llama3-8b.b1742.f16.gguf llama3-8b
python llama-cpp-b3286/convert-hf-to-gguf.py --outtype f16 --outfile llama3-8b/llama3-8b.b3286.f16.gguf llama3-8b
```

Perplexity evaluation script:
```python
import torch
from datasets import load_dataset
from fire import Fire
from tqdm import trange
from transformers import AutoModelForCausalLM as ModelImp
from transformers import AutoTokenizer as TkCls
from transformers import PreTrainedModel as ModelCls


@torch.inference_mode()
def evaluate_perplexity(model_id, gguf_file=None, seqlen=2048):
    model: ModelCls = ModelImp.from_pretrained(
        model_id,
        device_map="auto",
        gguf_file=gguf_file,
        torch_dtype=torch.float16,
    )

    tk = TkCls.from_pretrained(model_id, gguf_file=gguf_file)
    dataset = load_dataset("wikitext", "wikitext-2-raw-v1", split="test")

    input_ids = list()
    for item in dataset:
        text = item["text"] + "\n"
        tokens = tk.encode(text, add_special_tokens=False)
        input_ids.extend(tokens)

    data_size = len(input_ids) // seqlen
    input_ids = input_ids[: data_size * seqlen]

    input_ids = torch.LongTensor(input_ids).view(data_size, seqlen)
    bos_token = torch.full((data_size, 1), tk.bos_token_id, dtype=torch.int64)
    input_ids = torch.concat((bos_token, input_ids), dim=1)
    input_ids = input_ids.to(model.device)

    batch_size = 1
    nlls = list()
    with trange(0, data_size, batch_size) as prog:
        for i in prog:
            batch = input_ids[i : i + batch_size]
            outputs = model(batch, labels=batch)
            nlls.append(outputs.loss)
            ppl = torch.exp(torch.stack(nlls).mean())
            prog.desc = f"ppl: {ppl:.4f}"


if __name__ == "__main__":
    Fire(evaluate_perplexity)
```

Evaluate perplexity:
```sh
python eval-ppl.py llama3-8b  # ppl: 6.2941
python eval-ppl.py llama3-8b llama3-8b.b3286.f16.gguf  # ppl: 1801.3766
python eval-ppl.py llama3-8b llama3-8b.b1742.f16.gguf  # ValueError: Trying to set a tensor of shape torch.Size([128256, 4096]) in "weight" (which has shape torch.Size([32000, 4096])), this look incorrect.
```

I've modified some integration code to make the f16 conversion work:

```python
# transformers/integrations/ggml.py
GGML_TYPES = {
    "F32": 0,
    "F16": 1,
    # ...
}

def load_dequant_gguf_tensor(shape, ggml_type, data):
    if ggml_type == GGML_TYPES["F32"]:
        values = data
    elif ggml_type == GGML_TYPES["F16"]:
        values = data
    # ...
```

### Expected behavior

The PPL of llama3 converted back from gguf is as high as 1801.3766, which is significantly different from the original model. However, the same steps do not cause this issue on TinyLlama.

Therefore, I referred to the upload time of [TheBloke/TinyLlama-1.1B-Chat-v1.0-GGUF](https://huggingface.co/TheBloke/TinyLlama-1.1B-Chat-v1.0-GGUF) and chose the version of llama.cpp released on January 1st for the gguf conversion. The following error occurred:
```
Converting and de-quantizing GGUF tensors...: 100%|███████████████| 291/291 [00:48<00:00,  6.01it/s]
Traceback (most recent call last):
  File "/data2/Penut/Experiments/240703-HF-GGUF/eval-ppl.py", line 48, in <module>
    Fire(evaluate_perplexity)
  File "/data2/Penut/.miniconda/envs/HF-GGUF/lib/python3.11/site-packages/fire/core.py", line 143, in Fire
    component_trace = _Fire(component, args, parsed_flag_args, context, name)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data2/Penut/.miniconda/envs/HF-GGUF/lib/python3.11/site-packages/fire/core.py", line 477, in _Fire
    component, remaining_args = _CallAndUpdateTrace(
                                ^^^^^^^^^^^^^^^^^^^^
  File "/data2/Penut/.miniconda/envs/HF-GGUF/lib/python3.11/site-packages/fire/core.py", line 693, in _CallAndUpdateTrace
    component = fn(*varargs, **kwargs)
                ^^^^^^^^^^^^^^^^^^^^^^
  File "/data2/Penut/.miniconda/envs/HF-GGUF/lib/python3.11/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/data2/Penut/Experiments/240703-HF-GGUF/eval-ppl.py", line 12, in evaluate_perplexity
    model: ModelCls = ModelImp.from_pretrained(
                      ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data2/Penut/.miniconda/envs/HF-GGUF/lib/python3.11/site-packages/transformers/models/auto/auto_factory.py", line 564, in from_pretrained
    return model_class.from_pretrained(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data2/Penut/.miniconda/envs/HF-GGUF/lib/python3.11/site-packages/transformers/modeling_utils.py", line 3838, in from_pretrained
    ) = cls._load_pretrained_model(
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data2/Penut/.miniconda/envs/HF-GGUF/lib/python3.11/site-packages/transformers/modeling_utils.py", line 4227, in _load_pretrained_model
    error_msgs, offload_index, state_dict_index = _load_state_dict_into_meta_model(
                                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data2/Penut/.miniconda/envs/HF-GGUF/lib/python3.11/site-packages/transformers/modeling_utils.py", line 895, in _load_state_dict_into_meta_model
    set_module_tensor_to_device(model, param_name, param_device, **set_module_kwargs)
  File "/data2/Penut/.miniconda/envs/HF-GGUF/lib/python3.11/site-packages/accelerate/utils/modeling.py", line 358, in set_module_tensor_to_device
    raise ValueError(
ValueError: Trying to set a tensor of shape torch.Size([128256, 4096]) in "weight" (which has shape torch.Size([32000, 4096])), this look incorrect.
```

No matter which version of llama.cpp is used, the PPL is consistent on TinyLlama. However, I noticed that when using a newer version of llama.cpp, there is an additional step in HF: "Merges were not in the checkpoint, building merges on the fly." 

The TinyLlama part can be reproduced using the following steps:

```sh
# tinyllama
HF_HUB_ENABLE_HF_TRANSFER=True huggingface-cli download TinyLlama/TinyLlama-1.1B-Chat-v1.0 --local-dir tinyllama-1b
python llama-cpp-b1742/convert.py --outtype f16 --outfile tinyllama-1b/tinyllama-1b.b1742.fp16.gguf tinyllama-1b
python llama-cpp-b3286/convert-hf-to-gguf.py --outtype f16 --outfile tinyllama-1b/tinyllama-1b.b3286.fp16.gguf tinyllama-1b

# ppl: 8.0233
python eval-ppl.py tinyllama-1b
python eval-ppl.py tinyllama-1b tinyllama-1b.b3286.fp16.gguf
python eval-ppl.py tinyllama-1b tinyllama-1b.b1742.fp16.gguf
```

### PR DESCRIPTION
**Pull request by PenutChen**
# What does this PR do?

Fixes #31766. The permutation of `q_proj` and `k_proj` needs to consider both `num_attention_heads` and `num_key_value_heads`.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@SunMarc

### CODE REVIEW

**Comment by LysandreJik** [comment]
Thanks for your PR @PenutChen, we really appreciate it!

@SunMarc it'd be awesome if you can review this when you have a minute.

**Review by SunMarc — COMMENTED** [review]
Thanks for the deep dive @PenutChen Really good job ! Can you add a few tests since it was not working with a bigger llama ! If possible, we can test this on a q4 model to not take too much memory !

File: src/transformers/modeling_gguf_pytorch_utils.py [hunk 1]
```diff
@@ -163,3 +164,12 @@ def load_gguf_checkpoint(gguf_checkpoint_path, return_tensors=False):
...
     return parsed_parameters
+
+
+def reverse_hf_permute(weights: np.ndarray, n_head: int, n_head_kv: int) -> np.ndarray:
+    if n_head_kv is not None and n_head != n_head_kv:
+        n_head = n_head_kv
+
+    dim = weights.shape[0] // n_head // 2
+    w = weights.reshape(n_head, dim, 2, *weights.shape[1:])
+    return w.swapaxes(2, 1).reshape(weights.shape)
```
[the remark this answers is not in the document]
Comment (PenutChen) [reply src/transformers/modeling_gguf_pytorch_utils.py]: This is quite complex. 🤔  
I looked at the original [permutation](https://github.com/ggerganov/llama.cpp/blob/a38b884c6c4b0c256583acfaaabdf556c62fabea/convert_hf_to_gguf.py#L1402-L1408) and tried to reverse it. Maybe I should follow the comment, like [`dequantize_q4_k`](https://github.com/huggingface/transformers/blob/43ffb785c0b2f24948d2011883d40dccb609d341/src/transformers/integrations/ggml.py#L219-L223), which just has a reference link?

**Comment by PenutChen** [comment]
@SunMarc Thanks for the advice! I noticed that there is already a Q4 Llama3 test. This update will change the expected text. Should I just update the expected text?

File: src/transformers/modeling_gguf_pytorch_utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (SunMarc) [reply src/transformers/modeling_gguf_pytorch_utils.py]: Sound good to me !

**Review by SunMarc — APPROVED** [review]
Thanks a lot for iterating ! Could you are the perplexity of the model before and after for llama 3 ? If the llama 3 perpelxity was so high, it's a bit strange that it was able to pass the test / generate coherent text

**Comment by PenutChen** [comment]
The perplexity of Llama3 Q4 is 1713.8865 before the fix and 6.4626 after the fix. The model comes from NousResearch/Meta-Llama-3-8B-GGUF.

They still produce reasonable output if the sequence length is short. This might be because only the q_proj and k_proj weights are abnormal, and their weights are relatively low. Other weights are robust to a certain extent, which could be the cause. This is just my guess.

**Comment by SunMarc** [comment]
Thanks for the confirmation !

**Review by LysandreJik — APPROVED** [review]
Sorry for the delayed review, thanks for the PR @PenutChen!

**Pull request merged**